### PR TITLE
Ultralight EV1/NTAG new dump format supports with new GUI compiled

### PR DESCRIPTION
**Pull request by McEloff**
1. Ultralight/EV1/NTAG simulation now checks counters, tearing flags, signature from new dump format
2. New compiled GUI supported new dump format for Ultralight EV1/NTAG

### CODE REVIEW

File: Firmware/Chameleon-Mini/Application/MifareUltralight.c [hunk 1]
```diff
@@ -379,37 +383,48 @@ static uint16_t AppProcess(uint8_t* const Buffer, uint16_t ByteCount)
...
-                uint32_t Counter;
+                uint32_t Counter = 0;
                 /* Validation */
                 if (CounterId > CNT_MAX) {
                     Buffer[0] = NAK_INVALID_ARG;
                     return NAK_FRAME_SIZE;
                 }
                 /* Read the value out */
-                MemoryReadBlock(&Counter, (PageCount + CounterId) * MIFARE_ULTRALIGHT_PAGE_SIZE, MIFARE_ULTRALIGHT_PAGE_SIZE);
+                MemoryReadBlock(&Counter, (PageCount + CounterId) * MIFARE_ULTRALIGHT_PAGE_SIZE, 3);
```
Comment (iceman1001) [inline Firmware/Chameleon-Mini/Application/MifareUltralight.c]: Don't use numbers.  Use a define macro / const,  to increase maintainability.

File: Firmware/Chameleon-Mini/Application/MifareUltralight.c [hunk 2]
```diff
@@ -379,37 +383,48 @@ static uint16_t AppProcess(uint8_t* const Buffer, uint16_t ByteCount)
...
                 /* Add and check for overflow */
                 Counter += Addend;
                 if (Counter > CNT_MAX_VALUE) {
                     Buffer[0] = NAK_CTR_ERROR;
                     return NAK_FRAME_SIZE;
                 }
                 /* Update memory */
-                MemoryWriteBlock(&Counter, (PageCount + CounterId) * MIFARE_ULTRALIGHT_PAGE_SIZE, MIFARE_ULTRALIGHT_PAGE_SIZE);
+                MemoryWriteBlock(&Counter, (PageCount + CounterId) * MIFARE_ULTRALIGHT_PAGE_SIZE, 3);
```
Comment (iceman1001) [inline Firmware/Chameleon-Mini/Application/MifareUltralight.c]: same as above

File: Firmware/Chameleon-Mini/Application/MifareUltralight.c [hunk 3]
```diff
@@ -379,37 +383,48 @@ static uint16_t AppProcess(uint8_t* const Buffer, uint16_t ByteCount)
...
-            case CMD_READ_SIG:
-                /* Hardcoded response */
-                memset(Buffer, 0xCA, SIGNATURE_LENGTH);
+            case CMD_READ_SIG: {
+                /* Check new dump format and get signature from out of pages area (skip 3 pages of counters, 2 pages VERSION) */
+                MemoryReadBlock(Buffer, (PageCount + 5) * MIFARE_ULTRALIGHT_PAGE_SIZE, SIGNATURE_LENGTH);
+                if (*(uint32_t*)Buffer == 0) {
+                    /* Hardcoded response */
+                    memset(Buffer, 0xCA, SIGNATURE_LENGTH);
+                }
```
Comment (iceman1001) [inline Firmware/Chameleon-Mini/Application/MifareUltralight.c:418]: better detection of old vs new format will be needed.

File: Firmware/Chameleon-Mini/Application/MifareUltralight.c [hunk 4]
```diff
@@ -379,37 +383,48 @@ static uint16_t AppProcess(uint8_t* const Buffer, uint16_t ByteCount)
...
                 ISO14443AAppendCRCA(Buffer, SIGNATURE_LENGTH);
                 return (SIGNATURE_LENGTH + ISO14443A_CRCA_SIZE) * 8;
+            }
 
-            case CMD_CHECK_TEARING_EVENT:
+            case CMD_CHECK_TEARING_EVENT: {
+                uint8_t CounterId = Buffer[1];
+                /* Read tearing flag from counter pages of new dump format */
+                MemoryReadBlock(Buffer, (PageCount + CounterId) * MIFARE_ULTRALIGHT_PAGE_SIZE + 3, 1);
```
Comment (iceman1001) [inline Firmware/Chameleon-Mini/Application/MifareUltralight.c]: same as above

**Review by iceman1001 — CHANGES_REQUESTED** [review]
Nice addition of new MFU dumps format.
We don't add the binary of the GUI to this repo.  You will have to remove that one.

**Comment by iceman1001** [comment]
@shinhub  @bogiton   Have you guys tested this one? 
I am not interested in adding the win gui binaries on this repo but the functionality I want.

**Comment by securechicken** [comment]
Sorry, do not have any Ultralight application at hand for now...

**Comment by iceman1001** [comment]
@McEloff   I like the functionality  but we are not adding GUI binaries into this repo.  You will need to remove that part.

**Pull request merged**